### PR TITLE
Custom dialog boxes!

### PR DESCRIPTION
**Pull request by frankitox16**
This is an approach to custom dialog boxes and a way to be called by LUA functions (that will be later added). It uses an XML File to define elements that will be part of the content of the dialog box, yet i've got to decide which kind of elements should a custom dialog box have (like a "FurnitureSelect" element or buttons, or things like that).

I also added a dictionary in the DialogBoxManager so that lua functions can call dialog boxes by name (not implemented yet)

Right now if you test it, this just adds the dialog box to de canvas, and updates it's title, however the whole TODO would be:

- [x] Add lua actions for custom dialog box.
- [x] Add more content for dialog boxes.
- [x] Allow LUA to show custom dialog boxes.
- [x] Add callback for custom dialog boxes.
- [x] Implement DialogBoxResult in custom dialog boxes.

To see the example that I've added, ~~you have to pause the game, and under Canvas->DialogBoxes you should see a "Testing" dialog box. Enable it~~ Right click a Stockpile and you'll see a "Set Filter" button. Press it and the dialog box that was loaded from the StreamingAssets/UI/DialogBoxes folder will appear. If you want to play a little bit with it, I'm making a [Wiki](https://github.com/TeamPorcupine/ProjectPorcupine/wiki/How-does-it-work%3F#adding-external-moddable-ui-dialog-box-frankitox16) page to show how it works and how to add your custom ones.

Bonus! A little screenshot of the example:
image

### CODE REVIEW

**Comment by dusho** [comment]
any chance you can use De/serialization instead of XmlReader? We won't do much about code we have now, but if any future code can use deserialization, we can port it easily to other data language if needed..
Also I wouldn't use DB instead of DialogBox.. (DialogBox isn't that long word) may be bit confusing for all the people knowing and using DB for database stuff..

**Comment by frankitox16** [comment]
@dusho hmm I don't have much experience with de/serialization. Is it something related with that XmlSerializable tag some documents have? About the DB stuff, yes I'll change it to DialogBox when I get home.

**Comment by dusho** [comment]
it is actually much easier than all XmlReader magic, easy example here:
https://support.microsoft.com/en-us/kb/815813
with way to customize the output:
http://stackoverflow.com/questions/5937630/serialize-class-to-xml

File: Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Xml;
+using System.Xml.Serialization;
+using UnityEngine;
+
+public class DialogComponent
+{
+    [XmlAttribute("Type")]
```
Comment (dusho) [inline Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs]: according to guidelines here, we have elements starting with capitals and attributes are camelCase, so:
`[XmlAttribute("type")]`

File: Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    public string ObjectType;
+
+    [XmlElement]
+    public Vector3 position;
+
+    [XmlElement]
```
Comment (dusho) [inline Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs]: [XmlAttribute("Data")]

File: Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+using System.Xml;
+using System.Xml.Serialization;
+using UnityEngine;
+
+public class DialogComponent
+{
+    [XmlAttribute("Type")]
+    public string ObjectType;
+
+    [XmlElement]
```
Comment (dusho) [inline Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs]: [XmlAttribute("Position")]
shame here that Unity doesn't serialize this with x,y,z as attributes. Though use can create wrapper class, adjust the format then have it return Unitys Vector3..

File: Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs (hunk 2 above)
Comment (dusho) [inline Assets/Scripts/Models/DialogBoxInformation/DialogComponent.cs]: ah ye, sorry.. copy-paste failed me, just wanted to put there capitals
[XmlElement("Data")]

File: Assets/Scripts/UI/DialogBox/LUA/DialogBoxLua.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    public void CancelButtonClicked()
+    {
+        Result = DialogBoxResult.Cancel;
+        events.Trigger("OnClosed", this, Result);
+        CloseDialog();
+    }
+
+    public void OkButtonClicked()
```
Comment (bjubes) [inline Assets/Scripts/UI/DialogBox/LUA/DialogBoxLua.cs]: you call the OnClosed trigger in each __ButtonClicked method, then call closeDialog in those methods, which calls the same trigger again. Remove those OnClosed trigger calls in the specific buttons but leave the call in the CloseDialog() Method

File: Assets/Scripts/UI/DialogBox/PromptOrInfo/DialogBoxResult.cs [hunk 5]
```diff
@@ -10,5 +10,6 @@ public enum DialogBoxResult
 {
     Yes,
     No,
-    Cancel
+    Cancel,
+    OK
```
Comment (bjubes) [inline Assets/Scripts/UI/DialogBox/PromptOrInfo/DialogBoxResult.cs]: OK -> Okay

File: Assets/Scripts/UI/DialogBox/LUA/DialogBoxLua.cs (hunk 4 above)
Comment (frankitox16) [inline Assets/Scripts/UI/DialogBox/LUA/DialogBoxLua.cs]: Whoops... begginer mistake

**Comment by frankitox16** [comment]
Okay, almost finished. Would like to add some more "components" like images or Inventory selection, but I'm tempted to leave it as it is, and allow people to add their stuff, as this isn't exactly a v0.1 feature.

**Comment by TomMalbran** [comment]
I don't like the use of Lua in the names. Dialogs could be just Xml or later we could decide not to use Lua. The name shouldn't mention the implementation.

File: Assets/Scripts/UI/DialogBox/LUA/DialogBoxLua.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        protected set
+        {
+            title = value;
+            transform.GetChild(0).GetChild(0).GetComponentInChildren<Text>().text = value;
+        }
+    }
+    
+    public override void ShowDialog()
+    {
+        base.ShowDialog();
```
Comment (bjubes) [inline Assets/Scripts/UI/DialogBox/LUA/DialogBoxLua.cs]: Suggestion: I would like a way to reference show dialog in addition to  close dialog from lua. that seems like a very likely thing to need, say if you wanted to have a sound play when a dialog opens.

**Comment by frankitox16** [comment]
@bjubes I've added the ShowDialogByName method and made it available for it's use on LUA

**Comment by bjubes** [comment]
Now that this is no longer WIP, can you specify how to test this? I don't see the test dialog you metioned in your OP so is there a way to open that dialog? or did you implement an existing dialog using the new lua way?

**Comment by frankitox16** [comment]
@bjubes Whoops! Should've added it in my description when i edited it. Anyway, if you pause the game and look in the dialog boxes container in the canvas, you'd see a "Testing" dialog box, which was loaded by the script. I'm in the process of adding how this works in the wiki, but knowing this you should be able to test it. 
The XML and lua files are in the StreamingAssets folder, under UI/DialogBoxes

File: Assets/Scripts/Models/DialogBoxInformation/DialogBoxLuaInformation.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+// ====================================================
+#endregion
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Xml.Serialization;
+using UnityEngine;
+
+public class DialogBoxLuaInformation
```
Comment (TomMalbran) [inline Assets/Scripts/Models/DialogBoxInformation/DialogBoxLuaInformation.cs]: Don't name the classes with Lua. The implementation shouldn't be part of the name. Same goes for all the other names that have Lua, like the prefab, and the DialogBoxLua.

Comment (bjubes) [inline Assets/Scripts/Models/DialogBoxInformation/DialogBoxLuaInformation.cs]: Maybe use the word mod instead, so ModDialogBox

Comment (frankitox16) [inline Assets/Scripts/Models/DialogBoxInformation/DialogBoxLuaInformation.cs]: I think I've got it. Please tell me if you find any more issues.

**Comment by bjubes** [comment]
a few things i noticed:
- the buttons on the dialog are different than the other dialogs (im basing that off of the dialogs dynamically created in code for confirming quit, creating a new world, etc...)
- when i use this code:
``` 
<Title>Testing</Title>
  <Content>
  	<Text>this is sample text</Text>
  </Content>
  <Buttons>Yes</Buttons>
... and so on
```
in testDB.xml no text shows up.
-  there is no way (or i am missing it ) to interface with dialogBoxPromptOrInfo using lua to make generic dialog boxes. this would be helpful when a mod author doesn't want to define multiple xml files just to change the text inside a simple popup. this may be as easy as adding [MoonSharpUserData] to to the top of the dialogBoxPromptOrInfo class

**Comment by frankitox16** [comment]
@bjubes 
Ok, I'll be trying to add the interaction with dialog prompts. As for the dialog buttons, could you please add an image showing the difference? I don't seem to get it.

Text doesnt show up because the `<content>` tag requires a "type" argument to specify what kind of content it is ([reference](https://github.com/TeamPorcupine/ProjectPorcupine/wiki/How-does-it-work%3F#adding-external-moddable-ui-dialog-box-frankitox16)), so in order to add text, you'd add: 
```
...
<Title>Testing</Title>
<Content name="customName" type="Text">
<Position>
<x></x>
<y></y>
<z></z>
</Position>
<Data>this is sample text</Data>
</Content>
...
```
The bad thing you made me notice is that `<Content>` is a terrible name for that, as it's not descriptive enough.

**Comment by bjubes** [comment]
capture
i just noticed the button size is inconsitent. however, i don't know if there is a actual "standard' size in the project or if all the buttons are varying sizes

**Comment by Tranberry** [comment]
some style guidelines would be good imo.

**Comment by frankitox16** [comment]
@bjubes I've added the interface with not only "prompt or info" dialog box, but also with the others (however not all of them implement it yet) and corrected the buttons to match prompt's button height.

@Tranberry would you give me an example of the style guidelines that could be added? So I could get a better idea of what you mean...

Also, special thanks to @koosemose and @TomMalbran for the help with the Lua interface, you helped me a lot!

**Comment by bjubes** [comment]
Changes look good from the diff, but now there are merge conflicts. Once you resolve them change the title and ill test this out!

**Comment by bjubes** [comment]
Sorry, I just merged in a PR before checking this one out so there are conflicts again. Ill make sure to test this one next after the conflicts are fixed

File: Assets/Scripts/UI/DialogBox/DialogBoxManager.cs [hunk 8]
```diff
@@ -39,42 +39,66 @@ public void Awake()
...
+            dialogBoxQuests = tempGoObj.GetComponent<DialogBoxQuests>();
+            AddQuestList();
 
-        tempGoObj = CreateDialogGO("DB_LoadFile", "Load File");
-        dialogBoxLoadGame = tempGoObj.GetComponent<DialogBoxLoadGame>();
-        DialogBoxes["Load File"] = dialogBoxSaveGame;
+            AddMainMenuItems();
+        }else
+        {
+            tempGoObj = CreateDialogGO("DB_SaveFile", "Save File");
```
Comment (Ohda) [inline Assets/Scripts/UI/DialogBox/DialogBoxManager.cs]: There is a problem with your merge conflict here. You don't load the good GameObject on the good scenes.

Comment (frankitox16) [inline Assets/Scripts/UI/DialogBox/DialogBoxManager.cs]: You mean the right* game objects on the right scenes, am I right? :sweat_smile: I found the difference with the actual code, the conflicts weren't exactly clear. I'll fix them right away

Comment (Ohda) [inline Assets/Scripts/UI/DialogBox/DialogBoxManager.cs]: French guys are bad with english. Sorry mate :)

**Comment by koosemose** [comment]
Looks mostly good. One minor nitpick, in the test dialog when you choose yes and it brings up the PromptOrInfo dialog, it continues closing (which also leaves the PromptOrInfo dialog as nonmodal), which is probably the right thing in most scenarios, however in the specific case of an "Are You Sure", I think the proper behavior would be to delay closing until a response is received from PromptOrInfo, and either continue closing on Yes or just not close on No. 

Is the system currently capable of having behavior like this and it just isn't wired up for the test case or is it not currently possible to do that. If it isn't that would be a nice addition if it isn't too much extra code (if it is a lot it could be done in a follow-up PR).

**Comment by frankitox16** [comment]
@koosemose Now that I've added listeners to the buttons, we're able to keep the DialogBox before closing the prompt.

**Comment by koosemose** [comment]
So that I know what to expect for testing, is the test made to do this, or is it just that the capability is generally there?

**Comment by frankitox16** [comment]
The "filter" test is just a way to prove the way it works, and how you'd modify it. However, that filter dialog may become the dialog that will set stockpile filters, when I figure out a proper way to select inventory/set priorities. As for the listeners, are for every modded dialog box, not for all of them (eg: the settings menu one)

**Comment by koosemose** [comment]
I mean the behavior of the dialog staying open during the "Are you sure" Dialog

**Comment by frankitox16** [comment]
That's why I changed the listeners. Before they would trigger an event AND close the dialog. Now they just trigger an event, so the closing of the dialog should be handled inside lua. Long story short: the prompt dialog will be displayed while the dialog is still shown, you can get a result from the prompt and act in consecuence, and in order to close the dialog, you must do it from lua

**Comment by koosemose** [comment]
Alright, just wanted to make sure my expectations of how things will work is correct, to minimize time wasted if I misunderstand how it's intended to perform and think it's not performing as it should. :)

**Comment by koosemose** [comment]
Looks good, but you've got some stylecop issues, once those are fixed we should be ready for merging (assuming of course nothing breaks unexpectedly)

**Comment by koosemose** [comment]
Conflicted

**Comment by koosemose** [comment]
Whoops wrong PR

**Comment by frankitox16** [comment]
StyleCop fixed. All clean

**Comment by koosemose** [comment]
You seem to have missed ModDialogBoxInformation as well.

**Comment by frankitox16** [comment]
done :sweat_smile:

**Comment by koosemose** [comment]
Looks good, stylecop clean, no new bugs rearing their nasty heads. Merged!

**Pull request merged**